assembly/src/parsers/ast/mod.rs

```diff
@@ -66,6 +69,53 @@ impl ProgramAst {
 
         Ok(ProgramAst { local_procs, body })
     }
+
+    /// Compiles the AST into a program that can be executed on Miden VM.
+    #[allow(unreachable_code, unused_variables)]
+    pub fn compile(
+        self,
+        context: &AssemblyContext,
+        cb_table: &mut CodeBlockTable,
+    ) -> Result<CodeBlock, AssemblyError> {
+        let Self {
+            local_procs,
+            mut body,
+        } = self;
+
+        // split the local procs into sets that should either be inlind or exported to the cb table
+        let (mut exported, inlined): (Vec<ProcedureAst>, Vec<ProcedureAst>) =
+            local_procs.into_iter().partition(|proc| proc.is_export);
+
+        // any call to an inlined procedure in an exported one should be inlined
+        exported.iter_mut().for_each(|_| todo!());
+
+        // generate a codeblock counterpart of the public procedures
+        let (exported_procs, exported_blocks): (Vec<_>, Vec<_>) = exported
+            .into_iter()
+            .map::<(ProcedureAst, CodeBlock), _>(|_| todo!())
+            .unzip();
+
+        // generate the digests of the exported procedures
+        //
+        // all the exported procs now share a common index with its code block and digest
+        // counterpart
+        let exported_digests: Vec<_> = exported_blocks.iter().map(CodeBlock::hash).collect();
+
+        // add public procedures to the cb table
+        exported_blocks
+            .into_iter()
+            .for_each(|block| cb_table.insert(block));
+
+        // inline all internal procedures in the body
+        body.iter_mut().for_each(|_| todo!());
```

Comment:
I am not sure I follow this. What we need to do is:
1. For each declared procedure, we need to convert the procedure into a `Procedure` struct and add it to the `AssemblyContext`. This way, every subsequent procedure will have access to MASTs of already parsed procedures.
2. Then we need to process the body, which will need to access `Procedure` structs from `AssemblyContexts`.

In the `AssemblyContext` these procedures will go under `local_procs`, which should be changed from a map to a vector. This is because in the new structure, local procedures are accesses via their index rather than their name (e.g., `ExecLocal`).

One other thought, we also need access to `ModuleProvider` here to resolve references to imported procedures. Though, this may be abstracted away under `AssemblyContext`.

Reply:
N/A; impl moved